src/main/java/com/google/devtools/build/lib/runtime/BlazeRuntime.java

```diff
@@ -1589,7 +1597,8 @@ public BlazeRuntime build() throws AbruptExitException {
           productName,
           serverBuilder.getBuildEventArtifactUploaderMap(),
           serverBuilder.getAuthHeadersProvidersMap(),
-          serverBuilder.getRepositoryRemoteExecutorFactory());
+          serverBuilder.getRepositoryRemoteExecutorFactory(),
+          serverBuilder.getDownloaderFactory());
```

Comment:
This looks wrong. This means it's not possible to have different Downloader factories for different builds in the same client. Shouldn't this be part of the CommandEnvironment instead?

Reply:
I don't understand the question, but handling of the remote downloader should match that of the other remote capabilities (caching and execution). The calls to `getRepositoryRemoteExecutorFactory()` and `getDownloaderFactory()` belong together.

If they belong together _somewhere else_, could we do that separately? I'm very out of my depth when it comes to this level of the Bazel client architecture.

Reply:
I think what ulf means is, assuming you're familiar with the client-server setup, the DownloaderFactory instance (and thus implementation) is tied to the lifetime of the server. It cannot be changed across different commands to the same server. I think this is fine since it's no worse than the current setup for other remote stuff, and there's only one impl.